crypto/bio/bss_dgram.c

```diff
@@ -1061,28 +1061,42 @@ static void translate_msg_win(BIO *b, WSAMSG *mh, WSABUF *iov,
 # if M_METHOD == M_METHOD_RECVMMSG || M_METHOD == M_METHOD_RECVMSG
 /* Translates a BIO_MSG to a msghdr and iovec. */
 static void translate_msg(BIO *b, struct msghdr *mh, struct iovec *iov,
-                          unsigned char *control, BIO_MSG *msg)
+                          unsigned char *control, BIO_MSG *msg, int connected)
 {
+
     iov->iov_base = msg->data;
     iov->iov_len  = msg->data_len;
 
-    /* macOS requires msg_namelen be 0 if msg_name is NULL */
-    mh->msg_name = msg->peer != NULL ? &msg->peer->sa : NULL;
-    if (msg->peer != NULL && dgram_get_sock_family(b) == AF_INET)
-        mh->msg_namelen = sizeof(struct sockaddr_in);
+    if (connected == 0) {
+	/* macOS requires msg_namelen be 0 if msg_name is NULL */
+	mh->msg_name = msg->peer != NULL ? &msg->peer->sa : NULL;
+	if (msg->peer != NULL && dgram_get_sock_family(b) == AF_INET)
+	    mh->msg_namelen = sizeof(struct sockaddr_in);
 #  if OPENSSL_USE_IPV6
-    else if (msg->peer != NULL && dgram_get_sock_family(b) == AF_INET6)
-        mh->msg_namelen = sizeof(struct sockaddr_in6);
+	else if (msg->peer != NULL && dgram_get_sock_family(b) == AF_INET6)
+	    mh->msg_namelen = sizeof(struct sockaddr_in6);
 #  endif
-    else
-        mh->msg_namelen = 0;
+	else
+	    mh->msg_namelen = 0;
+    } else {
+	mh->msg_name = NULL;
+	mh->msg_namelen = 0;
+    }
 
     mh->msg_iov         = iov;
     mh->msg_iovlen      = 1;
     mh->msg_control     = msg->local != NULL ? control : NULL;
     mh->msg_controllen  = msg->local != NULL ? BIO_CMSG_ALLOC_LEN : 0;
     mh->msg_flags       = 0;
 }
+
+static int is_connected(BIO *b)
+{
+    struct sockaddr_storage ss;
+    socklen_t ss_len = sizeof(ss);
+
+    return (getpeername(b->num, (struct sockaddr *)&ss, &ss_len) == 0);
```

Comment:
I do not think we want to add a syscall to every `BIO_sendmmsg` invocation.

Whether a socket is connected is supposed to be tracked by BIO_CTRL_DGRAM_SET_CONNECTED (`data->connected`).